### ISSUE TITLE
[Egg Request]: Generic Eggs for using github deploy keys/ssh instead of PATs/https

### ISSUE THREAD

**Issue by steelskillet** [parkervcp/eggs#2808]
### Does this expand an already existing service

Yes

### Link to a game or software

https://docs.github.com/en/authentication/connecting-to-github-with-ssh/managing-deploy-keys#deploy-keys

### Links for the download

N/A

### Links for the install docs

https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent?platform=linux

**Comment by QuintenQVD0** [comment]
This can not work with pterodactyl as there is no way to pull / upload your own ssh key and their is no input in the install container to enter a ssh passphrase.

**Comment by parkervcp** [comment]
I meant to comment on this earlier.

It wouldn't really be feasible to have ssh keys.
1. You would need the private key in a place where the panel user can read it. Most likely a mount.
2. It would need to be owned by the user to use it on top of that.

**Comment by QuintenQVD0** [comment]
\+ it would need a wings modification as mounts are not present in the install container

**Comment by steelskillet** [comment]
You can specify ssh key locations and they can be stored in the server container. A generated key could then be retrieved via the file manager or sftp.

Basic procedure would be to have the install script generate the key pairs, setup ssh-agent to use that key, then pull from GitHub using ssh.
Updates/startup would then just be pulling from GitHub using ssh.

User side the server would install itself. There would then be a generated key pair in `home/container` in which the public key could be copied from.
There's no reason to be using mounts to external sources on the server.

**Comment by steelskillet** [comment]
@QuintenQVD0 It seems you misunderstood the implementation i wanted for this. I absolutely understand that key pairs outside of the container would be unfeasible at the moment however that's not the implementation I was hoping for.

The installation steps, in regards to pulling a repository from github, I would like to see for this egg are as follows:
1: generate ssh-key pair files and save to `/mnt/server` (or a similar sub-directory)
2: register the key pair with ssh-agent
3: attempt to clone the repository using ssh (it fails due to no pub key, that's fine and it should just continue)

after installation the panel user would be able to copy the public key from the service using the file manager or sftp and add it to the github repo or their own account.

The run steps for the egg for pulling from the repo would then be:
(0?): register the key pair with ssh-agent, unsure if this is needed.
1: clone or pull the repository using ssh (logic is basically `AUTO_UPDATE==1 && not contains .git` then clone, otherwise if contains .git then pull).

**Comment by QuintenQVD0** [comment]
> @QuintenQVD0 It seems you misunderstood the implementation i wanted for this. I absolutely understand that key pairs outside of the container would be unfeasible at the moment however that's not the implementation I was hoping for.
> 
> The installation steps, in regards to pulling a repository from github, I would like to see for this egg are as follows: 1: generate ssh-key pair files and save to `/mnt/server` (or a similar sub-directory) 2: register the key pair with ssh-agent 3: attempt to clone the repository using ssh (it fails due to no pub key, that's fine and it should just continue)
> 
> after installation the panel user would be able to copy the public key from the service using the file manager or sftp and add it to the github repo or their own account.
> 
> The run steps for the egg for pulling from the repo would then be: (0?): register the key pair with ssh-agent, unsure if this is needed. 1: clone or pull the repository using ssh (logic is basically `AUTO_UPDATE==1 && not contains .git` then clone, otherwise if contains .git then pull).

I wont be implementing this as I see no need for it, if you are anyone else want to do it fine for me then I will review the PR but I do not have time for this or see the need for it.

**Comment by steelskillet** [comment]
ssh keys are much easier to manage than PATs and can be tied to specifically to a repo rather than to a user. This is a important distinction for private repositories owned by an organization rather than an individual.
With using PATs and https it's currently impossible to use a private repository of an organization with PATs disabled with the current eggs.

And to my knowledge at least, using ssh is the more modern access method.

**Comment by parkervcp** [comment]
Your idea is to create a new ssh key and let the install fail to then copy the ssh key manually to github so it is then allowed to clone the repo.

A Personal Access Token (PAT from here forward) is generated before hand and can be set as a variable so the install works for the initial install.

SSH keys are much older than PATs and are not limited in what they can do. They are essentially the user with full access. PATs have permissions down to a single repo if you want.

**Comment by steelskillet** [comment]
For the egg, all we care about is cloning and reading the repository, we don't need to say, edit issues or PRs, so we don't exactly care or want that extra functionality provided by PATs. While not an ideal solution to basically let the installation fail (it still installs say, the yolk and python for example), it also allows for similar functionality with minimal changes made to the egg and it's process. Unless if we want the user themselves to generate the ssh key and put that as a configuration option, that however puts extra work and needed technical knowledge on the user.

The thing that separates ssh keys over PATs is that they can also be tied directly to a repository and not a user. This is needed in specific scenarios where PAT access for users is disabled for a private repository owned by an organization.

ssh keys have always been more secure than PATs as you never transmit the sensitive portion of the key pair that authenticates you as you over the wire, unlike with PATs. There's a reason ssh keys have been around for a long time. Fine-grain PATs are also still in a BETA state while classic tokens don't have individual repository restrictions.

TLDR: There scenarios where PATs are unable to access a repository for read access where an ssh key could. ssh keys are more secure and more fitted for deployment scenarios, like with eggs and pterodactyl due to being long-lived and more flexible about where they can be used for read access.

**Comment by parkervcp** [comment]
So you are talking about a [deploy key](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/managing-deploy-keys#deploy-keys) not a standard ssh key. A deploy key is very different than a standard ssh key. You still need to fail the install first then have a user add the ssh key to the repo.

If you remove a user from a repo their ssh deploy key still has access where PATs lose access to any repos that the user doesn't have access too. I don't see how an ssh key is more secure here.

I am keeping this closed as PATs can do everything you are currently requesting. They get access to public repos and other repos the user has access to as well. Even if fine grained PATs are in beta it's still something usable and can be tried out by users.

**Comment by steelskillet** [comment]
1: yes, that is one way to use it. But like i stated earlier as well, ssh keys can be used to authenticate to git repos outside of github as well, which you can't do with PATs.

Deploy keys work exactly the same as a standard ssh key, only difference is read and possibly write access is owned by a single repository, rather than owned by a user. otherwise, you could still generate the key pair beforehand on your own machine and place into the installation config and repo before attempting installation if you wanted (and if there was such config to use).

2: with using PATs, if a user leaves the organization, they could break deployment because their PAT no longer has read access. where as a deploy key wouldn't. what we are really trying to authenticate in this instance with running pterodactyl is the container not a specific user, ssh keys make more sense for this.

3: if an organization has PATs disabled (which they can), the only way to access a private repo owned by the organization would be ssh keys. Also PATs cannot be used to authenticate to outside of github.
